Implement the rendezvous protocol

thomaseizinger opened the pull request:
This is an attempt of implementing the rendezvous protocol. This also fully includes https://github.com/libp2p/rust-libp2p/pull/2081. The first commits are pretty messy but later down the track, they are quite atomic in case you prefer looking at things patch-by-patch.

## A couple of design decisions that are worth mentioning:

- In-memory storage only. I see this more as a temporary thing to get an MVP out the door. Whether we want persistent storage in the form of an SQL database can be discussed in a follow-up. For example, I am not fully sold that we need it due to the ephemeral nature of registrations.
- ~PoW is implemented as per the [drafted extension](https://github.com/libp2p/specs/pull/334) to the protocol. The extension itself is still very much WIP, so any feedback is highly appreciated.~
- ~PoW is done on a separate thread. I did this because depending on the requested difficulty and machine power, performing the PoW can take several seconds and doing it as part of a task would essentially be blocking. I am not super happy with spawning threads inside the handler though because it makes the code  non-portable to platforms like wasm which would be a shame because I think all the rest is.~
- ~We only support one inbound and one outbound substream to a specific peer at any point.~
- The test suite and substream management may appear a bit over-engineered but I wanted to experiment with some designs that might be generalizable for other protocols to make implementing protocols easier.

## Things that I think we need to discuss before this can be merged:

1. ~How should we handle keep-alive in the handler? This is a part of rust-libp2p that I am not super familiar with so any guidance here would be much appreciated :)~
1. ~Usage of tokio for timers. I know that rust-libp2p usually prefers async-std so we will need to come to an agreement of how this can be solved.~
1. ~Do we want to have different modes for the behaviour such that a client could disallow registrations of other peers or should every node always also act as a rendezvous node?~
1. ~Is the split between behaviour and handler good? Currently, the `ProtocolsHandler` is quite "smart" and handles all the validation ~(including PoW!)~ and substream state management. This felt right to me because it allows the behaviour to deal with decision-making only.~
1. Do we want an "auto-renew" behaviour for registrations? As a behaviour that registers with a rendezvous point, we have all the necessary information to automatically renew the registration as it comes close to expiry. This might be something that all implementations want and hence should be provided together with the protocol. I am thinking of extending the `register` API with an argument `Renew::{Never, AfterRelativeElapsed(Precent)}` where `AfterRelativeElapsed` describes the relative time elapsed of the effective ttl after which we should renew (i.e. 50% of TTL elapsed).
1. Registering with a rendezvous node currently assumes that we want to register all of our external addresses as reported by other behaviours. This is quite opinionated but I like how it makes the interface very simple! I think we should change this so that we the user can choose between registering external addresses or providing a fixed list. Possibly in a similar way as the auto-renew functionality suggested above where the user can pass an enum `Addresses::{AllExternal, Given(Vec<Multiaddress>)}`

For (5) and (6), consider using a builder for a `NewRegistration` struct to avoid explosion of parameters.

## Things we are still planning on doing (in order of importance):

- [x] Transfer TODOs / FIXMEs from the code to this list
- [x] Add examples for a client, a server and an application that registers itself with a rendezvous point on startup
- [x] Fix panic on subsequent substream being injected inside `inject_fully_negotiated_*`
- [x] Fix panic on of "invalid handler state" in `inject_event`
- [x] Rip out PoW (see https://github.com/libp2p/specs/pull/334#issuecomment-865171882)
- [x] Implement `limit` for DISCOVER requests
- [x] Emit event for expired registrations
- [x] Validate that incoming registration is from requesting node
- [x] Have repeated registrations refresh existing registrations instead of duplicating them
- [x] Replace `tokio::time` with `wasm-timer`
- [x] Implement `connection_keep_alive` logic
- [x] Handle bad cookie being sent from client
- [x] Check the spec again and implement things like max namespace length etc (consider doing this by creating new-types for `namespace` and `ttl` to be more descriptive)
- [ ] Implement basic connection handling
	- [ ] allow kicking off a registration whilst not being connected yet, similar to `libp2p-request-response`
	- [x] take `Discovered` responses and make the information available via `addresses_of_peer` to help with connection management
- [x] ~Audit dependencies and remove stuff like `thiserror`~
- [x] Client-only mode to disallow incoming connections
	- [ ] Clarify with spec how clients should handle this (disallowing upgrades vs sending `NotAvailable` error?)
- [ ] Add proptests for converting between wire messages and codec messages
- [x] ~Allow the user to filter the list of external addresses being registered in `Rendezvous::register`~
- [ ] Extend spec with max message size?
- [x] Clarify why we need the `id` field in `Unregister` message (https://github.com/libp2p/specs/issues/335)
- [x] Extend `PeerRegistered` with registration object
- [x] Clarify with spec if there is an `UnregisterResponse` (https://github.com/libp2p/specs/issues/335)

burdges commented:
Why is there PoW in a rendezvous protocol?  That sounds pretty useless.

thomaseizinger commented:
> Why is there PoW in a rendezvous protocol? That sounds pretty useless.

@burdges It is marked in the spec as TBD for the final spec.

There isn't any rationale mentioned but I think the idea is to avoid spamming of rendezvous points. Obviously, the logic for how much difficulty is required would need to be reasonably sophisticated to make that actually work.

For example, if a rendezvous point determines the required difficulty by looking at the peers PeerId, external address and the number of existing registrations for that peer, spam can be mitigated to some extent.

Rendezvous points can be a criticial piece of infrastructure for a decentralized network. As such, I don't think it is unreasonable to take measures that reduce likelyhood of them being attacked through DoS.

I've created a PR for extending the spec here: https://github.com/libp2p/specs/pull/334
There are a lot of things to be discussed in regards to the implementation already. Do you mind taking the discussion on the general idea of PoW to the spec PR?

burdges commented:
I think DDoS adversaries are botnets, so they often have massive untapped CPU, GPU, and even hard drive.  I'd think honest peers cannot quickly solve a challenge within a difficulty level that presents an obstacle for adversaries.  

You should probably make applications choose their rendezvous points and simply document that they should do so carefully using their own deterministic randomness.  We'll eventually use a rendezvous-like protocol in polkadot, but we'll dictate the rendezvous point by VRF.  Tor also dictates introduction points also via deterministic randomness, not a VRF per se but is satisfies their security properties, including being anonymous to the rendezvous point.

thomaseizinger commented:
> I think DDoS adversaries are botnets, so they often have massive untapped CPU, GPU, and even hard drive. I'd think honest peers cannot quickly solve a challenge within a difficulty level that presents an obstacle for adversaries.
> 
> You should probably make applications choose their rendezvous points and simply document that they should do so carefully using their own deterministic randomness. We'll eventually use a rendezvous-like protocol in polkadot, but we'll dictate the rendezvous point by VRF. Tor also dictates introduction points also via deterministic randomness, not a VRF per se but is satisfies their security properties, including being anonymous to the rendezvous point.

Those are really interesting points, thank you for sharing them! I've opened [an issue](https://github.com/libp2p/specs/issues/337) on the specs repository to continue the discussion.

mxinden commented:
Thanks for the extensive work and the detailed description @thomaseizinger.

I am not sure how to interpret the fact that this is opened as a draft. Let me know once you want a (high-level) review here.

thomaseizinger commented:
I've marked it as ready for review. To start with, an initial high-level review would be good!

If you have something that needs changing across module-boundaries or otherwise architecturally significant, I would prefer to get that sorted before we dive into details :)

mxinden commented on protocols/rendezvous/src/handler.rs:
```diff
@@ -0,0 +1,520 @@
+    }
+
+    fn connection_keep_alive(&self) -> KeepAlive {
+        KeepAlive::Yes
```
To keep things simple, would the following work for now?

On each activity, e.g. inbound or outbound request, set `self.keep_alive = KeepAlive::Until(Instant::now() + ten_seconds)`. In `connection_keep_alive` return `self.keep_alive`.

mxinden reviewed (commented):
Did a high-level review.

Thank you for the diligent implementation thus far!

thomaseizinger commented on protocols/rendezvous/src/behaviour.rs:
```diff
@@ -0,0 +1,846 @@
+            }
+        };
+
+        self.events.extend(new_events);
```
I had plans to extract this mapping into a separate function and unit test it but never got around to it :)

I find it somewhat cleaner to produce a list of new events as part of handling the incoming one instead of each branch causing side-effects. I agree though that it is somewhat error prone. Maybe extracting a function is good enough to enforce the return value of a `Vec` of events?

thomaseizinger replied:
> I had plans to extract this mapping into a separate function

Done as part of https://github.com/libp2p/rust-libp2p/pull/2107/commits/8f46361d90c95ea340bb95c5edd7aa60324e2a75.

thomaseizinger commented on protocols/rendezvous/src/handler.rs:
```diff
@@ -0,0 +1,520 @@
+                outbound,
+            ),
+            (event, inbound, outbound) => {
+                log::warn!(
```
Yes this is intentional. If this is hit, we have a bug in one of the substream state machines.

mxinden replied:
> If this is hit, we have a bug in one of the substream state machines.

In that case I would suggest using a `debug_assert!` or even an `assert!`. State mismatch seems critical enough. In my opinion logs go unnoticed to often.

thomaseizinger commented on protocols/rendezvous/src/handler.rs:
```diff
@@ -0,0 +1,520 @@
+use void::Void;
+
+pub struct RendezvousHandler {
+    outbound: SubstreamState<Outbound>,
+    outbound_history: MessageHistory,
```
We don't actually use long-lived streams, it is a stream per {REGISTER,UNREGISTER,DISCOVER}-RESPONSE pair that is being closed again after the pair of messages is done.

Yeah I don't like the complexity of this very much but I figured this is the most correct way of going about the implementation :)

Note to self: Continue working on https://github.com/libp2p/rust-libp2p/discussions/2078#discussioncomment-839032 to make protocol implementations less verbose.

thomaseizinger commented on protocols/rendezvous/src/behaviour.rs:
```diff
@@ -0,0 +1,846 @@
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
```
Usually, I only use `thiserror` in applications and don't force the `syn` dependency onto users of a library, esp. for trivial stuff like `thiserror`. If you are okay with leaving it in, I am happy though. Less work for us :)

thomaseizinger commented on protocols/rendezvous/src/behaviour.rs:
```diff
@@ -0,0 +1,846 @@
+use tokio::time::sleep;
+use uuid::Uuid;
+
+pub struct Rendezvous {
```
Yep that is an interesting aspect.

I am worried that it will introduce duplicated code because I think that even if a node only acts as a client, they are expected to return `E_UNAVAILABLE` if another node wants to register with us. Hence, we cannot just use things like `upgrade::DeniedUpgrade` to completely disallow inbound substreams.

Consequently, a `RendezvousPointBehaviour` - which would handle inbound substreams - would have to accept every substream regardless and have a config switch to deny them using above error code. To be spec compliant, every node would therefore always include both behaviours anyway, even if one of them is disabled (side note: does this composition even work? Can we have two behaviours that handle the same protocol?)

Thinking about this more, I think I do see your point in that even if both behaviours are always included, splitting the implementation into two roles might make understanding it easier.

We talked about this internally (cc @da-kami @rishflab) but I may have disregarded the idea back then too fast :)
It is quite some work though so we might not jump on it right away :)

thomaseizinger commented:
@mxinden This is ready for a review!

mxinden commented:
> @mxinden This is ready for a review!

Great. I will review it next week. Sorry, I won't find the time this week.

thomaseizinger commented:
> > @mxinden This is ready for a review!
> 
> Great. I will review it next week. Sorry, I won't find the time this week.

No rush, it is a lot of code 😅

mxinden reviewed (commented):
_I have not yet fully reviewed the whole pull request, though I think there is use in me publishing my comments thus far early. Feel free to ask me to finish first, before following up._

Again, thank you for the diligent implementation and especially thank you for showcasing your suggestions detailed in  https://github.com/libp2p/rust-libp2p/discussions/2078.

On the rendezvous specifcs, see comments.

On the new abstraction:

I do see the utility of introducing a simplified abstraction to `ProtocolsHandler`. I find the name `SubstreamHandler` great. I am not yet sure I am in favor of the `advance` pattern, simply as it is yet another pattern (next to the functionally equivalent `Stream` trait) to learn for newcomers. As mentioned in the comments, I need to give it more thought. Food for thought: Have you considered a `SubstreamHandler` to implement `Sink` instead of `inject_event` and `Stream` instead of `advance`?

From where I stand today, I would favor:

- Renaming `ProtocolsHandler` to `ConnectionHandler` (thus consistent with `SubstreamHandler`).
- Introduce `SubstreamHandler` and `SubstreamConnectionHandler` (which would implement `ConnectionHandler`).
- Update simple protocols like `libp2p-ping` to use `SubstreamHandler` (please don't start this yet).
- Writing extensive documentation on which one to use when.

thomaseizinger commented on protocols/rendezvous/tests/harness/mod.rs:
```diff
@@ -0,0 +1,206 @@
+        .boxed();
+
+    SwarmBuilder::new(transport, behaviour_fn(peer_id, identity), peer_id)
+        .executor(Box::new(GlobalSpawnTokioExecutor))
```
Nope it isn't, this is a copy-paste from code that was created prior to https://github.com/libp2p/rust-libp2p/pull/1686 :)

Will remove!

thomaseizinger commented on protocols/rendezvous/src/substream_handler.rs:
```diff
@@ -0,0 +1,463 @@
+    }
+}
+
+fn poll_substreams<TId, TSubstream, TError, TOutEvent>(
```
It could be but only as an associated function because it is being called with both `&mut self.inbound_substreams` and `&mut self.outbound_substreams`. Hence, I didn't see an advantage of putting it within an `impl SubstreamProtocolsHandler`.

mxinden replied:
Might be worth a comment :)

thomaseizinger commented on protocols/rendezvous/src/substream_handler.rs:
```diff
@@ -0,0 +1,463 @@
+
+    fn new(substream: NegotiatedSubstream, info: Self::OpenInfo) -> Self;
+    fn inject_event(self, event: Self::InEvent) -> Self;
+    fn advance(self, cx: &mut Context<'_>) -> Result<Next<Self, Self::OutEvent>, Self::Error>;
```
The core aspect of this design is actually that it takes `self` by value instead of as a reference.

Looking at implementations of substream state machines, it is common that you actually need ownership to advance the state machine which results in `std::mem::replace` being used all the time. By providing an interface that gives you `self` directly instead of `&mut self`, I hoped to centralize the use of `std::mem::replace` into one place (`SubstreamProtocolsHandler`) and at the same time, it allows for a neat way of "shutting" down a `SubstreamHandler` why dropping the internal stream.

thomaseizinger replied:
From your review post:

> Have you considered a SubstreamHandler to implement Sink instead of inject_event and Stream instead of advance?

Above mentioned reason is also why I think this is not going to work as well because all `Sink`/`Stream` methods take `&mut self`.

thomaseizinger commented on protocols/rendezvous/src/codec.rs:
```diff
@@ -0,0 +1,624 @@
+
+    pub fn from_wire_encoding(mut bytes: Vec<u8>) -> Result<Self, InvalidCookie> {
+        // check length early to avoid panic during slicing
+        if bytes.len() < 16 {
```
There isn't any. As far as I can see, cookies are opaque bytes to the client and their format is not specified. For us, the cookie is a `Uuid` (which is always 16 bytes), concatenated with the namespace (if provided during discover). Hence, we need to have at least 16 bytes here to avoid panics during slicing.

mxinden replied:
Do I understand correctly that an implementation using a 15 byte cookie would thus not be compatible with this implementation?

mxinden replied:
Ah, right, thanks. I somehow assumed other entities in a network to treat this in some other way than an opaque list of bytes.

thomaseizinger commented on protocols/rendezvous/src/behaviour.rs:
```diff
@@ -0,0 +1,950 @@
+}
+
+#[derive(Debug, Eq, PartialEq, Hash, Copy, Clone)]
+struct RegistrationId(Uuid);
```
No particular reason, we can change it to a `u64` if you prefer.

thomaseizinger commented on protocols/rendezvous/src/behaviour.rs:
```diff
@@ -0,0 +1,950 @@
+                ));
+            }
+
+            let action = match PeerRecord::new(self.keypair.clone(), external_addresses) {
```
I think that depends partly on how we decide in regards to point (6) mentioned in the PR description.

1. We can move ownership of the keypair into `Swarm` AND always imply creating a `signed_peer_record` from all external addresses, i.e.:

```rust
impl PollParameters {
	fn signed_peer_record(&self) -> PeerRecord;
}
```

2. We can move ownership of the keypair into `Swarm` but allow the caller to specify, which addresses to sign i.e.:

```rust
impl PollParameters {
	fn signed_peer_record(&self, addresses: impl Iterator<Item = Multiaddr>) -> PeerRecord;
}
```

I think we should definitely move ownership of the keypair into `Swarm`. `libp2p-gossipsub` would also benefit from this I believe (cc @AgeManning).

In regards to controlling which addresses can be signed, I haven't formed a strong opinion yet but I am bit wary of being too opinionated here. Having `signed_peer_record` and `external_addresses` compose well seems like the better design, i.e. something like this:

```rust
fn poll(&mut self, cx: ..., poll_params: impl PollParameters) {
	let record = poll_params.signed_peer_record(poll_params.external_addresses());
}

```

thomaseizinger commented on protocols/rendezvous/src/behaviour.rs:
```diff
@@ -0,0 +1,950 @@
+    /// Hold addresses of all peers that we have discovered so far.
+    ///
+    /// Storing these internally allows us to assist the [`libp2p_swarm::Swarm`] in dialing by returning addresses from [`NetworkBehaviour::addresses_of_peer`].
+    discovered_peers: HashMap<(PeerId, Namespace), Vec<Multiaddr>>,
```
Never at the moment, I guess we also want to track the returned TTL of the registration and garbage-collect them ...

thomaseizinger commented:
@mxinden Sorry for the delay on this, things have been very busy at work and outside of it as well. I hope to get to working in the feedback on this sometime next week.

mxinden commented:
> Is there anyway how I can combine two `ProtocolHandlers` that define logic for the same protocol but one only inbound and one outbound?

Would be following be an option?

- Combine the two handlers via a `ProtocolsHandlerSelect` (or an `IntoProtocolsHandlerSelect`).
- The outbound-only protocol would return a `DeniedUpgrade` on `IntoProtocolsHandler::inbound_protocol` and `ProtocolsHandler::listen_protocol`, which returns an `iter::empty()` on `<DeniedUpgrade as UpgradeInfo>::protocol_info`.
- The inbound-only protocol would never emit a `ProtocolsHandlerEvent::OutboundSubstreamRequest`.

thomaseizinger commented:
> > Is there anyway how I can combine two `ProtocolHandlers` that define logic for the same protocol but one only inbound and one outbound?
> 
> Would the following be an option?
> 
>     * Combine the two handlers via a `ProtocolsHandlerSelect` (or an `IntoProtocolsHandlerSelect`).
> 
>     * The outbound-only protocol would return a `DeniedUpgrade` on `IntoProtocolsHandler::inbound_protocol` and `ProtocolsHandler::listen_protocol`, which returns an `iter::empty()` on `<DeniedUpgrade as UpgradeInfo>::protocol_info`.
> 
>     * The inbound-only protocol would never emit a `ProtocolsHandlerEvent::OutboundSubstreamRequest`.

Well, I tried to do that using my new fancy `SubstreamProtocolsHandler` but it is not quite working as expected :(
I am unsure where the bug is.

Can you confirm that using the same protocol string in two different handlers works if one is used for inbound and one for outbound?

mxinden commented:
> Can you confirm that using the same protocol string in two different handlers works if one is used for inbound and one for outbound?

I can't confirm, given that, if I am not mistaken, none of our existing protocols cover this case. E.g. https://github.com/libp2p/rust-libp2p/pull/2059 uses two protocols (hop and stop), thus never running into this case.

That said, nothing comes to my mind that would prevent two handlers using the same protocol string.

mxinden commented on protocols/rendezvous/src/client.rs:
```diff
@@ -0,0 +1,294 @@
+                handle_outbound_event(message, peer_id, &mut self.discovered_peers)
+            }
+            handler::OutboundOutEvent::InboundError { .. } => {
+                // TODO: log errors and close connection?
```
:+1:

thomaseizinger commented:
> > Can you confirm that using the same protocol string in two different handlers works if one is used for inbound and one for outbound?
> 
> I can't confirm, given that, if I am not mistaken, none of our existing protocols cover this case. E.g. #2059 uses two protocols (hop and stop), thus never running into this case.
> 
> That said, nothing comes to my mind that would prevent two handlers using the same protocol string.

Okay, I'll see if I can find the bug in my implementation then :)

mxinden reviewed (commented):
> What do you see as critical before we can merge this?

Couple of comments below. Nothing big.

Following up on https://github.com/libp2p/rust-libp2p/pull/2107#discussion_r673241208, whether the swarm should hand out signed peer records, or whether each behaviour should sign their own and thus own a private key:

I would suggest we proceed here as is, with `libp2p-rendezvous` signing the record. Once another behaviour has the need for signed peer records, I think we should consider once more moving the logic to `libp2p-swarm` or `libp2p-core`.

thomaseizinger commented on protocols/rendezvous/Cargo.toml:
```diff
@@ -0,0 +1,40 @@
+rand = "0.8"
+wasm-timer = "0.2"
+
+[target.'cfg(target_arch = "wasm32")'.dependencies]
+uuid = { version = "0.8", features = ["wasm-bindgen"] }
```
I guess so yes!

thomaseizinger commented:
@mxinden This is ready for another review!

I think I addressed all the feedback and managed to get everything working with a client/server split. Also dropped UUID to make wasm support easier.

mxinden reviewed (approved):
This looks good to me. Thanks for the overall work and the recent follow-ups.

Would you mind adding an entry to `core/CHANGELOG.md`, initialize `protocols/rendezvous/CHANGELOG.md` and reference it in `./CHANGELOG.md`?

thomaseizinger commented:
> This looks good to me. Thanks for the overall work and the recent follow-ups.
> 
> Would you mind adding an entry to `core/CHANGELOG.md`, initialize `protocols/rendezvous/CHANGELOG.md` and reference it in `./CHANGELOG.md`?

Done!

mxinden commented on core/src/network/event.rs:
```diff
@@ -109,7 +109,7 @@ where
     ///
     ///   * it encounters an error, which includes the connection being
-    ///     closed by the remote. In this case `error` is `Some`.
+    ///     closed by the remote. In this case `error` is `ome`.
```
```suggestion
    ///     closed by the remote. In this case `error` is `Some`.
```

mxinden replied:
Fixed in https://github.com/libp2p/rust-libp2p/pull/2218. Unfortunately I can not apply suggestions on a branch on the comit-network repository.
